Display subnetworks

Bachibouzouk opened the pull request:
Example of use

(You need to have oemof-network installed from https://github.com/oemof/oemof-network/pull/53)

from oemof.visio import ESGraphRenderer
gr = ESGraphRenderer(
    energy_system=energysystem,
    filepath="depth_example.png"
)
gr.render(max_depth=2)

[depth_example.zip](https://github.com/user-attachments/files/22115433/depth_example.zip)

With max_depth set to 3
<img width="863" height="672" alt="depth_example_3" src="https://github.com/user-attachments/assets/0d79e34a-8f62-42a4-90f3-deb60ef8f76c" />


With max_depth set to 2
<img width="843" height="505" alt="depth_example_2" src="https://github.com/user-attachments/assets/77d98fcc-3c37-4d4a-ad1d-f4563808eb06" />


With max_depth set to 1

<img width="471" height="405" alt="depth_example_1" src="https://github.com/user-attachments/assets/c9e1d779-adf5-4e36-ad25-19effdd3bdfc" />

Bachibouzouk committed:
Move code into a method

The loop over the energy system nodes

Bachibouzouk committed:
Add a method to draw subnetworks

Bachibouzouk committed:
Sort the energy_system node to draw the subnetworks first

Bachibouzouk committed:
Avoid duplicate connection within a SubNetwork

Bachibouzouk committed:
Make sure the label is always string rep for the drawing

Bachibouzouk committed:
Improve autocleanup of digraph

Bachibouzouk committed:
Simplify sorting of atomic nodes

Bachibouzouk committed:
Add depth information about the nodes

Bachibouzouk committed:
Adapt the code with the new Node depth property from oemof-network

Now the network depth information is directly available from the node as
property

Bachibouzouk committed:
Add a function to list the external connections of a SubNetwork

Bachibouzouk committed:
Introduce max_depth attribute

Bachibouzouk committed:
Make visio compatible with earlier oemof-network versions

Bachibouzouk committed:
Lint with black

Bachibouzouk committed:
Working when max_depth is one lower than the deepest component

Bachibouzouk committed:
Using the depth to decide to draw a component or not

Component links of depth gap larger than 1, need to be collected by the
extern_connections fonction

Bachibouzouk committed:
Draw all connection to the last allowed component before the max depth

Bachibouzouk committed:
Allow the user to set the max_depth on render() call

Example of use

from oemof_visio import ESGraphRenderer
gr = ESGraphRenderer(
    energy_system=energysystem,
    filepath="depth_example.png"
)
gr.render(max_depth=2)

p-snft commented on src/oemof/visio/energy_system_graph.py:
```diff
@@ -28,6 +29,22 @@
+        pass
+
+    SUBNETWORK_MODULE = False
+    warnings.warn(
+        "Your oemof-network version doesn't support the SubNetwork, if "
+        "you want to work with SubNetworks, please update oemof-solph"
+    )
```
Do we need to warn here? If the users have an old version of oemof.network, they do not use `SubNetwork`s  should be fine without the warning.

Bachibouzouk replied:
We don't need to :)

Bachibouzouk replied:
I removed it now

Bachibouzouk commented on src/oemof/visio/energy_system_graph.py:
```diff
@@ -85,6 +102,34 @@ def fixed_width_text(text, char_num=10):
+        return None
+
+
+def extern_connections(nnode):
```
I see this as code which could be placed within `SubNetwork` class

Bachibouzouk committed:
Remove warning in case of oemof-network version without SubNetwork

Bachibouzouk committed:
Replace print statements by logging

Bachibouzouk committed:
Infer the max_depth if not provided

Bachibouzouk committed:
Remove unused code

Bachibouzouk committed:
Add a method to get the dot file

This can be used for testing purposes

from oemof.visio import ESGraphRenderer
gr = ESGraphRenderer(
    energy_system=energysystem
)

with open("ref_graph.dot") as f:
    ref_dot = f.read()

assert gr.source() == ref_dot

Where we would have saved the example energy_systems to the dot file
using

with open("ref_graph.dot", "w") as f:
    f.write(gr.source())

Bachibouzouk committed:
Lint with black

Bachibouzouk commented on src/oemof/visio/energy_system_graph.py:
```diff
@@ -325,30 +417,69 @@ def connect(self, a, b):
+        self._generate_graph(max_depth)
         self.dot.pipe(**kwargs)
 
+    def source(self, max_depth=None):
```
In order to use this to run tests one would need to test with different version of oemof-network I believe

Bachibouzouk committed:
Make the code back-compatible with oemof-solph 0.6.0

Bachibouzouk committed:
Fix display of folded SubNetwork with a depth gap of 2

Bachibouzouk committed:
Change default label text width

Bachibouzouk committed:
Adapt the code to changes in oemof-Network

Bachibouzouk committed:
Add test case for subnetwork

Currently the test is only ran manually to generate 3 pictures, we could
automate the tests in the future and compare generated dot files with
reference dot files

Bachibouzouk committed:
Update README with new fonctionnalities

Bachibouzouk committed:
Set the lowest depth from 1 to 0

p-snft commented on src/oemof/visio/energy_system_graph.py:
```diff
@@ -161,68 +195,118 @@ def __init__(
+
+        for nd in components_to_add:
+            # Back-compatibility with oemof-solph 0.6.0 and lower
+            if SUBNETWORK_MODULE is False:
+                if not hasattr(nd, "depth"):
+                    setattr(nd, "depth", 0)
```
In my opinion, `visio` should not alter the data that it means to display.

Bachibouzouk replied:
I agree we can just requier the latest solph version

Bachibouzouk replied:
Removed in f726c9aed629d53a0cc58afd7667b83cf3a9176f

p-snft commented on src/oemof/visio/energy_system_graph.py:
```diff
@@ -325,30 +409,92 @@ def connect(self, a, b):
-            a = fixed_width_text(a.label, char_num=self.txt_width)
+
+        # Back-compatibility with oemof-solph 0.6.0 and lower
+        if SUBNETWORK_MODULE is False:
+            for n in (a, b):
+                if not hasattr(n, "depth"):
+                    setattr(n, "depth", 0)
```
Same as above.

Bachibouzouk replied:
Removed in f726c9aed629d53a0cc58afd7667b83cf3a9176f

p-snft reviewed (commented):
You do a lot of stuff to achieve backward compatibility with past versions of `oemof.network`. I would rather require the current version and scrap the required workarounds.

Bachibouzouk commented on tests/oemof_model_subnetworks.py:
It is a test in the sense I could generate the 3 pictures quickly and test visually that they display what we expected from them. It could become a test with the .dot file, we could compare it with the generated dot similarly to the LP file in solph. I didn't have time to implement this, but I would rather do this in a subséquent PR

Bachibouzouk committed:
WIP

Bachibouzouk committed:
Remove SubNetwork and AtomicNode

p-snft commented on src/oemof/visio/energy_system_graph.py:
```diff
@@ -102,7 +95,7 @@ def extern_connections(nnode):
     ext_outputs = []
     for sn in nnode.subnodes:
-        if isinstance(sn, SubNetwork):
+        if len(sn.subnodes) > 0:
```
This should do:

```suggestion
        if sn.subnodes:
```

Bachibouzouk committed:
Implement suggestion to test if node has subnodes

Bachibouzouk committed:
Remove back-compatibility code

p-snft committed:
Move example to examples directory

This example was created to test things but not in the sense of
unit-testing.
